src/Dscp/DB/SQLite/Queries.hs

```diff
@@ -0,0 +1,347 @@
+           and  assignment_hash = ?
+    |]
+
+getGradesForCourseAssignments :: DBM m => StudentId -> CourseId -> m [PrivateTx]
+getGradesForCourseAssignments student course = do
+    query getGradesForCourseAssignmentsQuery (course, student)
+  where
+    getGradesForCourseAssignmentsQuery :: Query
+    getGradesForCourseAssignmentsQuery = [q|
+        select     Submissions.student_addr,
+                   Submissions.contents_hash,
+                   Assignments.course_id,
+                   Assignments.contents_hash,
+                   Assignments.desc,
+                   Submissions.signature
+                   grade
+                   time
+
+        from       Transactions
+
+        left join  Submissions
+               on  submission_hash = Submissions.hash
+
+        left join  Assignments
+               on  sssignment_hash = Assignments.hash
+              and  Assignments.course_id = ?
+
+        where      student_addr = ?
+    |]
+
+getStudentTransactions :: DBM m => StudentId -> m [PrivateTx]
+getStudentTransactions student = do
+    query getStudentTransactionsQuery (Only student)
+  where
+    getStudentTransactionsQuery :: Query
+    getStudentTransactionsQuery = [q|
+        select     Submissions.student_addr,
+                   Submissions.contents_hash,
+                   Assignments.course_id,
+                   Assignments.contents_hash,
+                   Assignments.desc,
+                   Submissions.signature
+                   grade
+                   time
+
+        from       Transactions
+
+        left join  Submissions
+               on  submission_hash = Submissions.hash
+
+        left join  Assignments
+               on  sssignment_hash = Assignments.hash
+
+        where      student_addr = ?
+    |]
+
+createCourse :: DBM m => CourseId -> Maybe Text -> m CourseId
+createCourse course desc = do
+    execute createCourseRequest (course, desc)
+    return course
+  where
+    createCourseRequest = [q|
+        insert into  Courses
+        values       (?, ?)
+    |]
+
+assert :: DBM m => m Bool -> DomainError -> m ()
+assert action message = do
+    yes <- action
+
+    unless yes $ do
+        throwIO message
+
+assertJust :: DBM m => m (Maybe a) -> DomainError -> m a
+assertJust action message = do
+    mb <- action
+
+    whenNothing mb $ do
+        throwIO message
+
+existsCourse :: DBM m => CourseId -> m Bool
+existsCourse course = do
+    exists existsCourseQuery (Only course)
+  where
+    existsCourseQuery = [q|
+        select  1
+        from    Courses
+        where   id = ?
+    |]
+
+existsStudent :: DBM m => StudentId -> m Bool
+existsStudent student = do
+    exists existsCourseQuery (Only student)
+  where
+    existsCourseQuery = [q|
+        select  1
+        from    Students
+        where   addr = ?
+    |]
+
+createStudent :: DBM m => StudentId -> m StudentId
+createStudent student = do
+    execute createStudentRequest (Only student)
+    return student
```

Comment:
Same.

Reply:
Here, in fact, the function should receive `Student` argument, not `StudentId`.

Reply:
Oh, looks like we don't have any students. I'll stop the functions to return the same data they receive, then.

Reply:
Done.